hacspec/src/machine_integers.rs

```diff
@@ -10,123 +10,208 @@
 
 use crate::prelude::*;
 
+macro_rules! implement_public_unsigned_mi {
+    ($t:ty,$bits:literal) => {
+        implement_public_mi!($t, $bits);
+        impl ModNumeric for $t {
+            /// (self - rhs) % n.
+            #[cfg_attr(feature="use_attributes", library(hacspec))]
+            fn sub_mod(self, rhs: Self, n: Self) -> Self {
+                println!("FOOOOOO {:?}, {:?}, {:?}", self, rhs, n);
+                let mut tmp = self;
+                while tmp < rhs {
+                    tmp += n;
+                }
+                println!("{:?} = {:?} - {:?}", tmp, rhs, tmp-rhs);
+                (tmp - rhs) % n
+            }
+            /// `(self + rhs) % n`
+            #[cfg_attr(feature="use_attributes", library(hacspec))]
+            fn add_mod(self, rhs: Self, n: Self) -> Self {
+                (self + rhs) % n
+            }
+            /// `(self * rhs) % n`
+            #[cfg_attr(feature="use_attributes", library(hacspec))]
+            fn mul_mod(self, rhs: Self, n: Self) -> Self{
+                (self * rhs) % n
+            }
+            /// `(self ^ exp) % n`
+            #[cfg_attr(feature="use_attributes", library(hacspec))]
+            fn pow_mod(self, _exp: Self, _n: Self) -> Self {
+                unimplemented!();
+            }
+            /// `self % n`
+            #[cfg_attr(feature="use_attributes", library(hacspec))]
+            fn modulo(self, n: Self) -> Self {
+                self % n
+            }
+        }
+    };
+}
+
+macro_rules! implement_public_signed_mi {
+    ($t:ty,$bits:literal) => {
+        implement_public_mi!($t, $bits);
+        impl ModNumeric for $t {
+            /// (self - rhs) % n.
+            #[cfg_attr(feature="use_attributes", library(hacspec))]
+            fn sub_mod(self, rhs: Self, n: Self) -> Self {
+                signed_mod(self as i128 - rhs as i128, n as i128) as $t
```

Comment:
Why do you need `i128` conversions everywhere here ?

Reply:
`signed_mod` is implemented only for `i128`. I guess we should implement it for all, then this gets a little nicer.